### ISSUE TITLE
make `TypeInterface::assert`, `TypeInterface::coerce` and `TypeInterface::assert` emit `function.alreadyNarrowedType`

### ISSUE THREAD

**Issue by bendavies** [php-standard-library/phpstan-extension#20]
Hi,

Similar to how `\assert(...)` works (https://phpstan.org/r/1a64fa56-fe04-47d5-a2ed-ef82e2206248), it would be nice if `TypeInterface::assert`, `TypeInterface::coerce` and `TypeInterface::assert` emitted the same errors, alerting the developer to when `Type` assertions are not necessary.

Thanks!

**Comment by ondrejmirtes** [comment]
Please show a specific PHP code using PSL for which you expect "always true" error to be reported.

**Comment by bendavies** [comment]
```php
<?php

use Psl\Type;

class Foo{}

Type\instance_of(Foo::class)->assert(new Foo());
Type\instance_of(Foo::class)->matches(new Foo());
Type\instance_of(Foo::class)->coerce(new Foo());
Type\int()->assert(1);
Type\int()->matches(1);
Type\int()->coerce(1);
// repeat for as many implementation of `TypeInterface` as possible
```

**Comment by ondrejmirtes** [comment]
Right now this one is reported:

```
 ------ --------------------------------------------------------------------------------------------
  Line   test.php
 ------ --------------------------------------------------------------------------------------------
  11     Call to method Psl\Type\TypeInterface<int>::matches() with 1 will always evaluate to true.
         🪪  method.alreadyNarrowedType
 ------ --------------------------------------------------------------------------------------------
```

**Comment by ondrejmirtes** [comment]
If I isolate the examples like this:

```php
function (): void {
	Type\instance_of(Foo::class)->assert(new Foo());
};

function (): void {
	Type\instance_of(Foo::class)->matches(new Foo());
};

function (): void {
	Type\instance_of(Foo::class)->coerce(new Foo());
};

function (): void {
	Type\int()->assert(1);
};

function (): void {
	Type\int()->matches(1);
};

function (): void {
	Type\int()->coerce(1);
};
```

both `matches` examples are reported:

```
 ------ ----------------------------------------------------------------------------------------------
  Line   test.php
 ------ ----------------------------------------------------------------------------------------------
  12     Call to method Psl\Type\TypeInterface<Foo>::matches() with Foo will always evaluate to true.
         🪪  method.alreadyNarrowedType
  24     Call to method Psl\Type\TypeInterface<int>::matches() with 1 will always evaluate to true.
         🪪  method.alreadyNarrowedType
 ------ ----------------------------------------------------------------------------------------------
```

**Comment by ondrejmirtes** [comment]
Not sure from the top of my head where it's getting lost. The logic is in https://github.com/phpstan/phpstan-src/blob/2.1.x/src/Rules/Comparison/ImpossibleCheckTypeHelper.php called from https://github.com/phpstan/phpstan-src/blob/2.1.x/src/Rules/Comparison/ImpossibleCheckTypeMethodCallRule.php#L90. Feel free to debug that.

**Comment by bendavies** [comment]
the isolation is interesting.
on its own, this will error:

```php
<?php

use Psl\Type;

class Foo{}

Type\instance_of(Foo::class)->matches(new Foo());
```

but this won't:

```php
<?php


use Psl\Type;

class Foo{}

Type\instance_of(Foo::class)->assert(new Foo());
Type\instance_of(Foo::class)->matches(new Foo());
```
